2 support zipping 2d arrays

hctrdev opened the pull request:
Closes #2

Commit:
#2: Add array extension to zip 2D arrays

Add a new extension method for arrays to allow zipping a two-dimensional
array. This works in the same way as the existing `ZipArray` for arrays
of a single dimension.

Commit:
#2: Add ZipArray on IOneBasedArray2D interface

Add the new ZipArray function on the `IOneBasedArray2D` interface and
implementation class.
Add test for the ZipArray function on the raw array.

ColmBhandal reviewed (changes requested):
Modify test to show that zipping works for arrays with non-matching row/column counts

Commit:
#2: Add new test for zipping arrays of different sizes

Add a new test to ensure the zip function correctly handles arrays of
different sizes. Add documentation on the methods.

ColmBhandal reviewed (changes requested):
Extend the array called second in the test to have an extra row- this demonstrates that both rows and columns that fall outside the intersection are allowed and will be ignored in the zipped result

Commit:
#2: Update test for 2D array zipping

Update the test for zipping 2D arrays to zip a 2x3 array with a 3x2 array
and show the result is a 2x2 array. This is instead of zipping a 2x3 and
a 2x2 array.
